nav2_route/src/graph_saver.cpp

```diff
@@ -0,0 +1,133 @@
+// limitations under the License.
+
+#include <memory>
+
+#include "nav2_route/graph_saver.hpp"
+#include "ament_index_cpp/get_package_share_directory.hpp"
+
+namespace nav2_route
+{
+
+GraphSaver::GraphSaver(
+  nav2_util::LifecycleNode::SharedPtr node,
+  std::shared_ptr<tf2_ros::Buffer> tf,
+  const std::string frame)
+: plugin_loader_("nav2_route", "nav2_route::GraphFileSaver"),
+  default_plugin_id_("GeoJsonGraphFileSaver")
+{
+  logger_ = node->get_logger();
+  tf_ = tf;
+  route_frame_ = frame;
+
+  nav2_util::declare_parameter_if_not_declared(
+    node, "graph_filepath", rclcpp::ParameterValue(std::string("")));
+  graph_filepath_ = node->get_parameter("graph_filepath").as_string();
+
+  // Default Graph Parser
+  const std::string default_plugin_type = "nav2_route::GeoJsonGraphFileSaver";
+  nav2_util::declare_parameter_if_not_declared(
+    node, "graph_file_saver", rclcpp::ParameterValue(default_plugin_id_));
+  auto graph_file_saver_id = node->get_parameter("graph_file_saver").as_string();
+  if (graph_file_saver_id == default_plugin_id_) {
+    nav2_util::declare_parameter_if_not_declared(
+      node, default_plugin_id_ + ".plugin", rclcpp::ParameterValue(default_plugin_type));
+  }
+
+  // Create graph file saver plugin
+  try {
+    plugin_type_ = nav2_util::get_plugin_type_param(node, graph_file_saver_id);
+    GraphFileSaver::Ptr graph_parser = plugin_loader_.createSharedInstance((plugin_type_));
+    RCLCPP_INFO(
+      logger_, "Created GraphFileSaver %s of type %s",
+      graph_file_saver_id.c_str(), plugin_type_.c_str());
+    graph_file_saver_ = std::move(graph_parser);
+    graph_file_saver_->configure(node);
+  } catch (pluginlib::PluginlibException & ex) {
+    RCLCPP_FATAL(
+      logger_,
+      "Failed to create GraphFileSaver. Exception: %s", ex.what());
+    throw ex;
+  }
+}
+
+bool GraphSaver::saveGraphToFile(
+  Graph & graph,
+  std::string filepath)
+{
+  if (filepath.empty() && !graph_filepath_.empty()) {
+    RCLCPP_DEBUG(
+      logger_, "The graph filepath was not provided. "
+      "Setting to %s", graph_filepath_.c_str());
+    filepath = graph_filepath_;
+  } else if (filepath.empty() && graph_filepath_.empty()) {
+    // No graph to try to save
+    return false;
+  }
+
+  RCLCPP_INFO(
+    logger_,
+    "Saving graph file from %s, by parser %s", filepath.c_str(), plugin_type_.c_str());
+
+  if (!graph_file_saver_->saveGraphToFile(graph, filepath)) {
+    return false;
+  }
+
+  if (!transformGraph(graph)) {
+    RCLCPP_WARN(
+      logger_,
+      "Failed to transform nodes graph file (%s) to %s!", filepath.c_str(), route_frame_.c_str());
+    return false;
+  }
+
+  return true;
+}
+
+bool GraphSaver::transformGraph(Graph & graph)
+{
+  std::unordered_map<std::string, tf2::Transform> cached_transforms;
+  for (auto & node : graph) {
+    std::string node_frame = node.coords.frame_id;
+    if (node_frame.empty() || node_frame == route_frame_) {
+      continue;
```

Comment:
Can we add a comment here explaining why we ignore this condition? It is obvious in this case when we look at the transform calculation below, but adding guiding comments generally make reading the code easier.